charms/istio-pilot/src/charm.py

```diff
@@ -251,6 +296,15 @@ def reconcile(self, event):
         # so that we can report them at the end.
         handled_errors = []
 
+        # Call upgrade_charm in case there are new configurations that affect the control plane
+        # only if the CNI configurations have been provided and have changed from a previous state
+        # This is useful when there is a missing configuration during the install process
+        try:
+            if self._cni_config_changed:
+                self.upgrade_charm(event)
+        except GenericCharmRuntimeError as err:
+            handled_errors.append(err)
+
```

Comment:
Since this can only trigger during a config change, can we move it to its own handler?  That way the logic for reconciling  the relations is separate from the reconfiguring of the istio daemon?  So something like

```python

class Operator(CharmBase):
  def __init__(self, *args):
    ...
    # Handle when istio deployment configuration is changed
    self.framework.observe(self.on.config_changed, self.update_istio_cni_configurations)

  def update_istio_cni_configurations(self, _):
    try:
      if self._cni_config_changed:
        self.upgrade_charm(event)
    except GenericCharmRuntimeError as err:
      handled_errors.append(err)
```

This'll help keep logic as encapsulated as we can

Reply:
Would that be possible though? We are already [observing config changed](https://github.com/canonical/istio-operators/blob/e574d9ecb2578b81e58338a521e2ad3ec6412345/charms/istio-pilot/src/charm.py#L133) events and tying them to the reconcile method.

Reply:
Definitely possible.  For example, the observability libs we use subscribe to the same events as some of our own handlers.  Juju just runs everything that subscribes in sequence (not sure if the order is defined)

But, given that we just talked today about splitting this charm into two (istio deployment and istio integrator), I say we just ignore this comment for now as soon it wont matter. I'll resolve this conversation